public/app/features/search/service/frontend.ts

```diff
@@ -29,20 +29,26 @@ export class FrontendSearcher implements GrafanaSearcher {
   }
 
   async getCache(kind?: string[]): Promise<FullResultCache> {
-    const key = kind ? kind.join(',') : '*';
-    let res = this.cache.get(key);
-    if (res) {
-      return Promise.resolve(res);
-    }
+    const key = kind ? kind.sort().join(',') : '*';
 
-    const v = await this.parent.search({
-      kind, // match the request
-      limit: 5000, // max for now
-    });
+    const cacheHit = this.cache.get(key);
+    if (cacheHit) {
+      return cacheHit;
+    }
 
-    res = new FullResultCache(v.view);
-    this.cache.set(key, res);
-    return res;
+    const resultPromise = this.parent
+      .search({
+        kind, // match the request
+        limit: 5000, // max for now
+      })
+      .then((res) => new FullResultCache(res.view))
+      .catch((err) => {
+        // TODO we need to refetch on error
```

Comment:
not sure if this would work, but what if the `catch` was removed and we checked for a rejected promise when there's a cache hit on line 35?

Reply:
we could, though I think it doesn't change much - right now the conversion from a failed promise to a result happens once, if we removed the `catch` it would happen N times

we need to add a retry/refresh logic that's going to touch this code anyway, but I think it could be done in a subsequent PR

Reply:
> right now the conversion from a failed promise to a result happens once

sorry, i don't have enough context about the larger flow, so i should have dug deeper before commenting. i guess i was thinking that this change is caching the empty result, so it wasn't clear how the user could issue the same query again without hitting the empty result

Reply:
updated the PR based on our slack convo!